### PR TITLE
Added commandline arguments to compile.sh

### PR DESCRIPTION
**Pull request by LogicParrot**
Closes #3211 and #3014

### CODE REVIEW

File: compile.sh [hunk 1]
```diff
@@ -1,41 +1,151 @@
...
+
+STATE_NEW=1
+if checkCuberiteDir; then # Check if we're in the Cuberite directory...
+	STATE_NEW=0
+elif [ -d cuberite ]; then # If there's a directory named "cuberite"...
+	cd cuberite
+	if checkCuberiteDir; then # Check if we're in the Cuberite directory...
+		STATE_NEW=0
+	else
+		errorOther "A directory named 'cuberite' already exists. Please run the script elsewhere or delete that directory."
```
Comment (LogicParrot) [inline compile.sh]: Make it more explicit that the `cuberite` directory has no actual git content.

File: compile.sh [hunk 2]
```diff
@@ -44,78 +154,96 @@ Compiling from source takes time, but it usually generates faster
...
-	# note that IFS is a TAB here!
-	echo "$PROGRAMS" | while IFS='	' read EXE_NAME PACKAGE_NAME __trash__; do
-		command -v $EXE_NAME > /dev/null 2> /dev/null || printf %s " $PACKAGE_NAME"
-	done
+	# Depdendency check.
+	checkPackages ()
+	{
+		# note that IFS is a TAB here!
+		echo "$PROGRAMS" | while IFS='	' read EXE_NAME PACKAGE_NAME __trash__; do
+			command -v $EXE_NAME > /dev/null 2> /dev/null || printf %s " $PACKAGE_NAME"
```
Comment (LogicParrot) [inline compile.sh]: That's a bit ugly looking. Can I clean it somehow?

File: compile.sh [hunk 3]
```diff
@@ -44,78 +154,96 @@ Compiling from source takes time, but it usually generates faster
...
+doDependencyCheck
 
-	# apt-get guide.
-	apt-get --help > /dev/null 2> /dev/null && \
-	missingDepsExit "sudo apt-get install $MISSING_PACKAGES"
 
-	# yum guide.
-	yum --help > /dev/null 2> /dev/null && \
-	missingDepsExit "sudo yum install $MISSING_PACKAGES"
+#=================== Choice: Branch Currently unused and simply skipped ===================
```
Comment (LogicParrot) [inline compile.sh]: `Choice: Branch (Currently unused and simply skipped)`

File: compile.sh [hunk 4]
```diff
@@ -138,101 +266,142 @@ You can choose between 3 branches:
...
+	* (D)Debug:   Compiles in debug mode.
+		      Makes your console and crashes more verbose.  
+		      A bit slower than Normal mode. If you plan to help
+		      development by reporting bugs, this is preferred.
 		
 
-Note that the script will connect to the internet in order to fetch
-code after this step. It will then compile your program.
-"
+	Note that the script will connect to the internet in order to fetch
```
Comment (LogicParrot) [inline compile.sh]: Move this note to thread part.

File: compile.sh [hunk 5]
```diff
@@ -138,101 +266,142 @@ You can choose between 3 branches:
...
+	echo "Chosen: " "$CHOICE_BUILDTYPE"
+elif [ ! -z "$CHOICE_BUILDTYPE" ]; then
+	CHOICE_BUILDTYPE="$DEFAULT_BUILDTYPE" # Non interactive mode with no buildtype specified
+fi
+
+#=================== Thread choice ===================
+
+chooseThreads()
+{
+	echo 1 #TODO
```
Comment (LogicParrot) [inline compile.sh]: todo

File: compile.sh [hunk 6]
```diff
@@ -138,101 +266,142 @@ You can choose between 3 branches:
...
+	echo 1 #TODO	
+}
+
+if [ $STATE_INTERACTIVE -eq 1 ]; then
+	printf %s "Enter the number of threads to use: "
+	read CHOICE_THREADS
+	echo "$CHOICE_THREADS"
+	if [ "$CHOICE_THREADS" == "" ] 2> /dev/null; then
+		CHOICE_THREADS=`chooseThreads`
+	elif [ "$CHOICE_THREADS" -gt 0 ] 2> /dev/null; then
```
Comment (LogicParrot) [inline compile.sh]: Briefly explain this "hack"

**Comment by bearbin** [comment]
@LogicParrot Any progress, this would be a good PR to merge.

File: compile.sh [hunk 7]
```diff
@@ -1,41 +1,151 @@
 #!/bin/sh
+set -e
```
Comment (LogicParrot) [inline compile.sh]: Why?

Comment (bearbin) [inline compile.sh]: If any of the commands return an error, the script will terminate instead of just continuing which is the default bash behaviour.

Why did you move it though?

Comment (LogicParrot) [inline compile.sh]: I experimented with something, but I cannot recall what it was.

**Comment by LogicParrot** [comment]
>@LogicParrot Any progress, this would be a good PR to merge. 

I'll be doing the final tests today and merge if it's ready.

**Comment by LogicParrot** [comment]
Can anyone test this on BSD's or Macs?

**Comment by Schwertspize** [comment]
Just a question: it may be useful to re-enable branch selection but with a definite warning. (if someone uses this flag, they know what it means afaik). Also we could then replace the CI scripts with something that must calls this script....
